apps/rush-lib/src/api/CommandLineConfiguration.ts

```diff
@@ -153,43 +167,72 @@ export class CommandLineConfiguration {
           );
         }
 
-        this.phases.set(phase.name, {
+        const processedPhase: IPhase = {
           ...phase,
           isSynthetic: false,
           logFilenameIdentifier: this._normalizeNameForLogFilenameIdentifiers(phase.name),
-          associatedParameters: new Set()
-        });
+          associatedParameters: new Set(),
+          index: -1,
+          phaseDependencies: {
+            self: new Set(),
+            upstream: new Set()
+          }
+        };
+
+        this.phases.set(phase.name, processedPhase);
       }
     }
 
+    // Resolve the string links
     for (const phase of this.phases.values()) {
-      if (phase.dependencies?.self) {
-        for (const dependencyName of phase.dependencies.self) {
+      const selfDependencies: string[] | undefined = phase.dependencies?.self;
+      const upstreamDependencies: string[] | undefined = phase.dependencies?.upstream;
+
+      if (selfDependencies) {
+        for (const dependencyName of selfDependencies) {
           const dependency: IPhase | undefined = this.phases.get(dependencyName);
           if (!dependency) {
             throw new Error(
               `In ${RushConstants.commandLineFilename}, in the phase "${phase.name}", the self ` +
                 `dependency phase "${dependencyName}" does not exist.`
             );
           }
+          phase.phaseDependencies.self.add(dependency);
         }
       }
 
-      if (phase.dependencies?.upstream) {
-        for (const dependency of phase.dependencies.upstream) {
-          if (!this.phases.has(dependency)) {
+      if (upstreamDependencies) {
+        for (const dependencyName of upstreamDependencies) {
+          const dependency: IPhase | undefined = this.phases.get(dependencyName);
+          if (!dependency) {
             throw new Error(
               `In ${RushConstants.commandLineFilename}, in the phase "${phase.name}", ` +
-                `the upstream dependency phase "${dependency}" does not exist.`
+                `the upstream dependency phase "${dependencyName}" does not exist.`
             );
           }
+          phase.phaseDependencies.upstream.add(dependency);
+        }
+      }
+    }
+
+    // Do the recursive stuff after the dependencies have been converted
+    const safePhases: Set<IPhase> = new Set();
+    for (const phase of this.phases.values()) {
+      const relatedPhaseSet: Set<IPhase> = new Set([phase]);
+      for (const relatedPhase of relatedPhaseSet) {
+        for (const dependency of relatedPhase.phaseDependencies.self) {
+          relatedPhaseSet.add(dependency);
+        }
+
+        for (const dependency of relatedPhase.phaseDependencies.upstream) {
+          relatedPhaseSet.add(dependency);
         }
       }
```

Comment:
`safePhases` stays empty through these loops. Is this code missing something?

Reply:
safePhases is mutated by _checkForPhaseSelfCycles and is used to skip redundant work while performing a depth-first search of the dependency graph.